File: crates/traderview-core/src/economic_value_added.rs
//! Economic Value Added (EVA) — the Stern Stewart measure of economic profit:
//! the after-tax operating profit a business earns above the cost of the capital
//! employed to earn it.
//!
//! ```text
//! NOPAT          = EBIT × (1 − tax rate)
//! capital charge = invested capital × WACC
//! EVA            = NOPAT − capital charge
//!                = (ROIC − WACC) × invested capital
//! ```
//!
//! Positive EVA means the business returns more than its capital costs (it
//! creates value); negative means it destroys value even if accounting profit is
//! positive. Distinct from `free-cash-flow` (cash generation), `dupont-roe`
//! (return on equity), and `wacc` (the discount rate this consumes).

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct EvaInput {
    /// Earnings before interest and taxes.
    pub ebit_usd: f64,
    pub tax_rate_pct: f64,
    /// Invested capital (debt + equity employed in operations).
    pub invested_capital_usd: f64,
    /// Weighted average cost of capital, percent.
    pub wacc_pct: f64,
    /// Revenue, for the EVA margin. 0 → margin omitted.
    #[serde(default)]
    pub revenue_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct EvaResult {
    /// EBIT × (1 − tax rate).
    pub nopat_usd: f64,
    /// Invested capital × WACC.
    pub capital_charge_usd: f64,
    pub eva_usd: f64,
    /// NOPAT / invested capital, percent. None if no capital.
    pub roic_pct: Option<f64>,
    /// ROIC − WACC, the value spread. None if ROIC is undefined.
    pub eva_spread_pct: Option<f64>,
    /// EVA / revenue, percent. None if no revenue.
    pub eva_margin_pct: Option<f64>,
    pub creates_value: bool,
}

pub fn analyze(input: &EvaInput) -> EvaResult {
    let nopat = input.ebit_usd * (1.0 - input.tax_rate_pct / 100.0);
    let capital_charge = input.invested_capital_usd * input.wacc_pct / 100.0;
    let eva = nopat - capital_charge;

    let roic = if input.invested_capital_usd != 0.0 {
        Some(nopat / input.invested_capital_usd * 100.0)
    } else {
        None
    };

    EvaResult {
        nopat_usd: nopat,
        capital_charge_usd: capital_charge,
        eva_usd: eva,
        roic_pct: roic,
        eva_spread_pct: roic.map(|r| r - input.wacc_pct),
        eva_margin_pct: if input.revenue_usd != 0.0 {
            Some(eva / input.revenue_usd * 100.0)
        } else {
            None
        },
        creates_value: eva > 0.0,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-2
    }

    fn base() -> EvaInput {
        EvaInput {
            ebit_usd: 1_000_000.0,
            tax_rate_pct: 25.0,
            invested_capital_usd: 5_000_000.0,
            wacc_pct: 8.0,
            revenue_usd: 4_000_000.0,
        }
    }

    #[test]
    fn nopat_and_charge() {
        let r = analyze(&base());
        // 1,000,000 × 0.75 = 750,000; 5,000,000 × 0.08 = 400,000.
        assert!(close(r.nopat_usd, 750_000.0));
        assert!(close(r.capital_charge_usd, 400_000.0));
    }

    #[test]
    fn eva_positive_creates_value() {
        let r = analyze(&base());
        assert!(close(r.eva_usd, 350_000.0));
        assert!(r.creates_value);
    }

    #[test]
    fn roic_and_spread() {
        let r = analyze(&base());
        // 750,000 / 5,000,000 = 15%; spread 15 − 8 = 7%.
        assert!(close(r.roic_pct.unwrap(), 15.0));
        assert!(close(r.eva_spread_pct.unwrap(), 7.0));
    }

    #[test]
    fn eva_margin() {
        let r = analyze(&base());
        // 350,000 / 4,000,000 = 8.75%.
        assert!(close(r.eva_margin_pct.unwrap(), 8.75));
    }

    #[test]
    fn spread_identity() {
        let r = analyze(&base());
        // EVA = (ROIC − WACC)/100 × invested capital.
        let from_spread = r.eva_spread_pct.unwrap() / 100.0 * 5_000_000.0;
        assert!(close(r.eva_usd, from_spread));
    }

    #[test]
    fn negative_eva_destroys_value() {
        let r = analyze(&EvaInput {
            ebit_usd: 200_000.0,
            ..base()
        });
        // nopat 150,000 − charge 400,000 = −250,000.
        assert!(close(r.eva_usd, -250_000.0));
        assert!(!r.creates_value);
        assert!(close(r.eva_spread_pct.unwrap(), -5.0));
    }

    #[test]
    fn breakeven_when_roic_equals_wacc() {
        // EBIT so NOPAT = capital charge (400,000 / 0.75).
        let r = analyze(&EvaInput {
            ebit_usd: 533_333.333333,
            ..base()
        });
        assert!(close(r.eva_usd, 0.0));
        assert!(close(r.eva_spread_pct.unwrap(), 0.0));
        assert!(!r.creates_value);
    }

    #[test]
    fn zero_capital_no_roic() {
        let r = analyze(&EvaInput {
            invested_capital_usd: 0.0,
            ..base()
        });
        assert!(r.roic_pct.is_none());
        assert!(r.eva_spread_pct.is_none());
        // No capital charge, so EVA equals NOPAT.
        assert!(close(r.eva_usd, 750_000.0));
    }

    #[test]
    fn no_revenue_no_margin() {
        let r = analyze(&EvaInput {
            revenue_usd: 0.0,
            ..base()
        });
        assert!(r.eva_margin_pct.is_none());
    }
}

File: frontend/js/views/economic_value_added.js
// Economic Value Added (EVA) — economic profit above the capital charge, via
// /calc/economic-value-added. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const money = (n) => (n == null ? '—' : (n < 0 ? '−$' : '$') + Math.abs(Number(n)).toLocaleString(undefined, { maximumFractionDigits: 0 }));
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%');
const VIEW = 'economic-value-added';
let lastReport = null;
let lastBody = null;

export async function renderEconomicValueAdded(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.eva.h1.title">// ECONOMIC VALUE ADDED</span></h1>
        <p class="muted small" data-i18n="view.eva.hint.intro">
            EVA is the after-tax operating profit a business earns above the cost of the capital it
            employs: NOPAT (EBIT after tax) minus invested capital times WACC. Positive EVA creates
            value; negative destroys it even when accounting profit is positive. Equivalently,
            (ROIC − WACC) × invested capital. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.eva.h2.inputs">The business</h2>
            <form id="eva-form" class="inline-form">
                <label><span data-i18n="view.eva.label.ebit">EBIT ($)</span>
                    <input type="number" step="1000" name="ebit_usd" value="1000000" required></label>
                <label><span data-i18n="view.eva.label.tax">Tax rate (%)</span>
                    <input type="number" step="0.1" min="0" name="tax_rate_pct" value="25" required></label>
                <label><span data-i18n="view.eva.label.capital">Invested capital ($)</span>
                    <input type="number" step="1000" min="0" name="invested_capital_usd" value="5000000" required></label>
                <label><span data-i18n="view.eva.label.wacc">WACC (%)</span>
                    <input type="number" step="0.1" min="0" name="wacc_pct" value="8" required></label>
                <label><span data-i18n="view.eva.label.revenue">Revenue ($, optional)</span>
                    <input type="number" step="1000" min="0" name="revenue_usd" value="4000000"></label>
            </form>
        </div>
        <div id="eva-tools" class="ce-toolbar"></div>
        <div id="eva-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#eva-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        return {
            ebit_usd: Number(fd.get('ebit_usd')) || 0,
            tax_rate_pct: Number(fd.get('tax_rate_pct')) || 0,
            invested_capital_usd: Number(fd.get('invested_capital_usd')) || 0,
            wacc_pct: Number(fd.get('wacc_pct')) || 0,
            revenue_usd: Number(fd.get('revenue_usd')) || 0,
        };
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcEconomicValueAdded(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.eva.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#eva-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'economic-value-added.csv' });
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['eva_usd', r.eva_usd],
        ['nopat_usd', r.nopat_usd],
        ['capital_charge_usd', r.capital_charge_usd],
        ['roic_pct', r.roic_pct],
        ['eva_spread_pct', r.eva_spread_pct],
    ];
}

function renderResult(mount, r) {
    const el = mount.querySelector('#eva-result');
    const cls = r.creates_value ? 'pos' : 'neg';
    // EVA decomposition: NOPAT minus capital charge equals economic value added.
    const chart = enh.svgBarChart([
        { label: 'NOPAT', value: r.nopat_usd },
        { label: 'Cap charge', value: -r.capital_charge_usd },
        { label: 'EVA', value: r.eva_usd },
    ]);
    const verdictKey = r.creates_value ? 'view.eva.verdict.creates' : 'view.eva.verdict.destroys';
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.eva.h2.result">The economic profit</h2>
            <div class="cards">
                <div class="card ${cls}"><div class="label" data-i18n="${verdictKey}">—</div>
                    <div class="value ${cls}">${money(r.eva_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.eva.card.roic">ROIC</div>
                    <div class="value">${pct(r.roic_pct)}</div></div>
                <div class="card"><div class="label" data-i18n="view.eva.card.spread">EVA spread</div>
                    <div class="value">${pct(r.eva_spread_pct)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.eva.row.nopat">NOPAT (EBIT after tax)</td><td>${money(r.nopat_usd)}</td></tr>
                    <tr><td data-i18n="view.eva.row.charge">Capital charge (capital × WACC)</td><td>${money(r.capital_charge_usd)}</td></tr>
                    <tr><td data-i18n="view.eva.row.roic">ROIC (NOPAT / capital)</td><td>${pct(r.roic_pct)}</td></tr>
                    <tr><td data-i18n="view.eva.row.spread">EVA spread (ROIC − WACC)</td><td>${pct(r.eva_spread_pct)}</td></tr>
                    <tr><td data-i18n="view.eva.row.margin">EVA margin</td><td>${pct(r.eva_margin_pct)}</td></tr>
                    <tr class="emph ${cls}"><td data-i18n="view.eva.row.eva">Economic Value Added</td><td>${money(r.eva_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}
